安装源代码包提示找不到 csproj.DotSettings 文件

安装源代码包提示

            C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(4643,5): error MSB3030: Could not copy the file "C:\Users\lindexi\.nuget\packages\lindexi.doubi.source\1.0.5-alpha179\src\lindexi.doubi.csproj.DotSettings" because it was not found.

--- a/build/Version.props
+++ b/build/Version.props
@@ -1,5 +1,5 @@
 <Project>
   <PropertyGroup>
-    <Version>0.1.19051-alpha</Version>
+    <Version>0.1.19097-alpha</Version>
   </PropertyGroup>
 </Project>

--- a/src/dotnetCampus.SourceYard/Context/PackagedProjectFile.cs
+++ b/src/dotnetCampus.SourceYard/Context/PackagedProjectFile.cs
@@ -80,7 +80,7 @@ namespace dotnetCampus.SourceYard.Context
 
                 // 如果文件放在项目文件上面，如 Foo\F1\F.csproj 引用文件 Foo\12.txt 那么此时就不添加 12.txt 的引用
                 // 原因是将 F.csproj 所在文件夹放在源代码包的 src 文件夹里面，因此 F1 文件夹是最上层文件夹
-                if (!fullPath.Contains(_projectFolder))
+                if (fullPath.IndexOf(_projectFolder, StringComparison.OrdinalIgnoreCase) < 0)
                 {
                     fileList.RemoveAt(i);
                     i--;
@@ -94,7 +94,7 @@ namespace dotnetCampus.SourceYard.Context
         {
             fileList.RemoveAll
             (
-                temp => IgnoreFolderList.Any(temp.StartsWith)
+                temp => IgnoreFolderList.Any(t => temp.StartsWith(t, StringComparison.OrdinalIgnoreCase))
             );
 
             fileList.RemoveAll(temp =>
@@ -102,13 +102,13 @@ namespace dotnetCampus.SourceYard.Context
                 var pathRoot = Path.GetPathRoot(temp);
                 if (!string.IsNullOrEmpty(pathRoot))
                 {
-                    return temp.StartsWith(pathRoot);
+                    return temp.StartsWith(pathRoot, StringComparison.OrdinalIgnoreCase);
                 }
 
                 return false;
             });
 
-            fileList.RemoveAll(temp => IgnoreFileEndList.Any(temp.EndsWith));
+            fileList.RemoveAll(temp => IgnoreFileEndList.Any(t => temp.EndsWith(t, StringComparison.OrdinalIgnoreCase)));
 
             return fileList;
         }

--- a/src/dotnetCampus.SourceYard/Context/PackagedProjectFile.cs
+++ b/src/dotnetCampus.SourceYard/Context/PackagedProjectFile.cs
@@ -1,52 +1,137 @@
-﻿namespace dotnetCampus.SourceYard.Context
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using dotnetCampus.SourceYard.Utils;
+
+namespace dotnetCampus.SourceYard.Context
 {
     /// <summary>
     /// 被打包项目的文件，包括编译的文件资源文件
     /// </summary>
     class PackagedProjectFile
     {
+        /// <inheritdoc />
+        public PackagedProjectFile(string compileFile, string resourceFile, string contentFile, string page, string applicationDefinition, string noneFile, string embeddedResource, string projectFolder)
+        {
+            _projectFolder = projectFolder;
+            ApplicationDefinition = applicationDefinition;
+
+            CompileFileList = GetFileList(compileFile);
+            ResourceFileList = GetFileList(resourceFile);
+            ContentFileList = GetFileList(contentFile);
+            PageList = GetFileList(page);
+            NoneFileList = GetFileList(noneFile);
+            EmbeddedResourceList = GetFileList(embeddedResource);
+        }
+
         /// <summary>
         /// 需要做源码包的项目的编译的文件
         /// </summary>
-        public string CompileFile { get; }
+        public IReadOnlyList<string> CompileFileList { get; }
 
         /// <summary>
         /// 需要做源码包的项目的资源文件
         /// </summary>
-        public string ResourceFile { get; }
+        public IReadOnlyList<string> ResourceFileList { get; }
 
         /// <summary>
         /// 需要做源码包项目的文件
         /// </summary>
-        public string ContentFile { get; }
+        public IReadOnlyList<string> ContentFileList { get; }
 
         /// <summary>
         /// 需要做源码包项目的页面
         /// </summary>
-        public string Page { get; }
+        public IReadOnlyList<string> PageList { get; }
 
         /// <summary>
         /// 嵌入文件
         /// </summary>
-        public string EmbeddedResource { get;}
+        public IReadOnlyList<string> EmbeddedResourceList { get; }
 
         /// <summary>
         /// 需要做源码包项目的文件
         /// </summary>
         private string ApplicationDefinition { get; }
 
-        public string NoneFile { get; }
+        public IReadOnlyList<string> NoneFileList { get; }
 
-        /// <inheritdoc />
-        public PackagedProjectFile(string compileFile, string resourceFile, string contentFile, string page, string applicationDefinition, string noneFile, string embeddedResource)
+        public IReadOnlyList<string> GetAllFile()
         {
-            CompileFile = compileFile;
-            ResourceFile = resourceFile;
-            ContentFile = contentFile;
-            Page = page;
-            ApplicationDefinition = applicationDefinition;
-            NoneFile = noneFile;
-            EmbeddedResource = embeddedResource;
+            return new CombineReadonlyList<string>(CompileFileList, ResourceFileList, ContentFileList, PageList,
+                NoneFileList, EmbeddedResourceList);
         }
+
+        private List<string> GetFileList(string file)
+        {
+            if (string.IsNullOrEmpty(file) || !File.Exists(file))
+            {
+                return new List<string>();
+            }
+
+            var fileList = File.ReadAllLines(file).ToList();
+
+            fileList = RemoveTempFile(fileList);
+
+            for (var i = 0; i < fileList.Count; i++)
+            {
+                var fullPath = Path.GetFullPath(fileList[i]);
+
+                // 如果文件放在项目文件上面，如 Foo\F1\F.csproj 引用文件 Foo\12.txt 那么此时就不添加 12.txt 的引用
+                // 原因是将 F.csproj 所在文件夹放在源代码包的 src 文件夹里面，因此 F1 文件夹是最上层文件夹
+                if (!fullPath.Contains(_projectFolder))
+                {
+                    fileList.RemoveAt(i);
+                    i--;
+                }
+            }
+
+            return fileList;
+        }
+
+        private List<string> RemoveTempFile(List<string> fileList)
+        {
+            fileList.RemoveAll
+            (
+                temp => IgnoreFolderList.Any(temp.StartsWith)
+            );
+
+            fileList.RemoveAll(temp =>
+            {
+                var pathRoot = Path.GetPathRoot(temp);
+                if (!string.IsNullOrEmpty(pathRoot))
+                {
+                    return temp.StartsWith(pathRoot);
+                }
+
+                return false;
+            });
+
+            fileList.RemoveAll(temp => IgnoreFileEndList.Any(temp.EndsWith));
+
+            return fileList;
+        }
+
+        /// <summary>
+        /// 项目文件夹
+        /// </summary>
+        private readonly string _projectFolder;
+
+        /// <summary>
+        /// 忽略的文件夹列表
+        /// </summary>
+        private static IList<string> IgnoreFolderList { get; } = new List<string>()
+        {
+            ".vs", "bin", "obj", ".git", "x64", "x86"
+        };
+
+        /// <summary>
+        /// 忽略的文件后缀列表
+        /// </summary>
+        private static IList<string> IgnoreFileEndList { get; } = new List<string>()
+        {
+            ".csproj.DotSettings", ".suo", ".user", ".sln.docstates", ".nupkg"
+        };
     }
 }

--- a/src/dotnetCampus.SourceYard/PackFlow/SourcePacker.cs
+++ b/src/dotnetCampus.SourceYard/PackFlow/SourcePacker.cs
@@ -1,5 +1,6 @@
 ﻿using System.Collections.Generic;
 using System.IO;
+using System.Linq;
 using dotnetCampus.SourceYard.Context;
 using dotnetCampus.SourceYard.Utils;
 
@@ -12,26 +13,9 @@ namespace dotnetCampus.SourceYard.PackFlow
             context.Logger.Message("开始复制源代码文件");
             var srcFolder = Path.Combine(context.PackingFolder, "src");
             context.Logger.Message("源代码临时复制文件夹 " + srcFolder);
-            FileSystem.IgnoreFileEndList = IgnoreFileEndList;
-            FileSystem.IgnoreFolderList = IgnoreFolderList;
-            FileSystem.CopyFolderContents(context.ProjectFolder, srcFolder, excludes: IgnoreFolderList);
+        
+            FileSystem.CopyFileList(context.ProjectFolder, srcFolder, context.PackagedProjectFile.GetAllFile().ToList());
             context.Logger.Message("复制源代码文件完成");
         }
-
-        /// <summary>
-        /// 忽略的文件夹列表
-        /// </summary>
-        private static IList<string> IgnoreFolderList { get; } = new List<string>()
-        {
-            ".vs", "bin", "obj", ".git", "x64", "x86"
-        };
-
-        /// <summary>
-        /// 忽略的文件后缀列表
-        /// </summary>
-        private static IList<string> IgnoreFileEndList { get; } = new List<string>()
-        {
-            ".csproj.DotSettings", ".suo", ".user", ".sln.docstates", ".nupkg"
-        };
     }
 }

--- a/src/dotnetCampus.SourceYard/Packer.cs
+++ b/src/dotnetCampus.SourceYard/Packer.cs
@@ -59,7 +59,8 @@ namespace dotnetCampus.SourceYard
                 embeddedResource: GetFile(embeddedResource),
                 page: GetFile(page),
                 applicationDefinition: GetFile(applicationDefinition),
-                noneFile: GetFile(noneFile)
+                noneFile: GetFile(noneFile),
+                projectFolder: Path.GetDirectoryName(_projectFile)
             );
 
             _packers = new IPackFlow[]

--- a/src/dotnetCampus.SourceYard/Utils/CombineReadonlyList.cs
+++ b/src/dotnetCampus.SourceYard/Utils/CombineReadonlyList.cs
@@ -1,0 +1,56 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace dotnetCampus.SourceYard.Utils
+{
+    public class CombineReadonlyList<T> : IReadOnlyList<T>
+    {
+        public CombineReadonlyList(params IReadOnlyList<T>[] source)
+        {
+            Source = source;
+        }
+
+        public IReadOnlyList<T>[] Source { get; }
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            foreach (var list in Source)
+            {
+                foreach (var temp in list)
+                {
+                    yield return temp;
+                }
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public int Count => Source.Sum(temp => temp.Count);
+
+        public T this[int index]
+        {
+            get
+            {
+                var n = index;
+                var source = Source;
+
+                foreach (var list in source)
+                {
+                    if (n < list.Count)
+                    {
+                        return list[n];
+                    }
+
+                    n -= list.Count;
+                }
+
+                throw new IndexOutOfRangeException();
+            }
+        }
+    }
+}

--- a/src/dotnetCampus.SourceYard/Utils/FileSystem.cs
+++ b/src/dotnetCampus.SourceYard/Utils/FileSystem.cs
@@ -8,6 +8,28 @@ namespace dotnetCampus.SourceYard.Utils
 {
     internal static class FileSystem
     {
+        /// <summary>
+        /// 复制文件列表
+        /// </summary>
+        /// <param name="sourceFolder"></param>
+        /// <param name="targetFolder"></param>
+        /// <param name="fileList"></param>
+        public static void CopyFileList(string sourceFolder, string targetFolder, List<string> fileList)
+        {
+            foreach (var file in fileList.Select(Path.GetFullPath))
+            {
+                var relativePath = MakeRelativePath(sourceFolder, file);
+                var targetFile = Path.GetFullPath(Path.Combine(targetFolder, relativePath));
+                var directory = Path.GetDirectoryName(targetFile);
+                if (directory != null && !Directory.Exists(directory))
+                {
+                    Directory.CreateDirectory(directory);
+                }
+
+                File.Copy(file, targetFile, true);
+            }
+        }
+
         internal static void CopyFolderContents(string sourceFolder, string targetFolder,
             Func<string, string> nameConverter = null, IList<string> excludes = null)
         {

--- a/src/dotnetCampus.SourceYard/Utils/ItemGroupElement.cs
+++ b/src/dotnetCampus.SourceYard/Utils/ItemGroupElement.cs
@@ -21,12 +21,12 @@ namespace dotnetCampus.SourceYard.Utils
         public (XElement itemGroupElement, XElement itemGroupElementOfXaml) GetItemGroup()
         {
             var contextPackagedProjectFile = _contextPackagedProjectFile;
-            var compileFileList = GetFileList(contextPackagedProjectFile.CompileFile);
-            var contentFileList = GetFileList(contextPackagedProjectFile.ContentFile);
-            var resourceFileList = GetFileList(contextPackagedProjectFile.ResourceFile);
-            var noneFileList = GetFileList(contextPackagedProjectFile.NoneFile);
-            var embeddedResource = GetFileList(contextPackagedProjectFile.EmbeddedResource);
-            var pageFileList = GetFileList(contextPackagedProjectFile.Page);
+            var compileFileList = contextPackagedProjectFile.CompileFileList;
+            var contentFileList = contextPackagedProjectFile.ContentFileList;
+            var resourceFileList = contextPackagedProjectFile.ResourceFileList;
+            var noneFileList = contextPackagedProjectFile.NoneFileList;
+            var embeddedResource = contextPackagedProjectFile.EmbeddedResourceList;
+            var pageFileList = contextPackagedProjectFile.PageList;
 
             var prefix = $"_{_packageGuid}";
 
@@ -53,7 +53,7 @@ namespace dotnetCampus.SourceYard.Utils
         private readonly bool _isVisible;
         private readonly string _packageGuid;
 
-        private List<XElement> XamlItemCompileFileToElement(List<string> compileFileList, string includingItemTypes,
+        private List<XElement> XamlItemCompileFileToElement(IEnumerable<string> compileFileList, string includingItemTypes,
             bool copyToOutputDirectory)
         {
             var elementList = new List<XElement>();
@@ -72,7 +72,7 @@ namespace dotnetCampus.SourceYard.Utils
             return elementList;
         }
 
-        private List<XElement> IncludingItemCompileFileToElement(List<string> compileFileList,
+        private List<XElement> IncludingItemCompileFileToElement(IReadOnlyList<string> compileFileList,
             string includingItemTypes, bool copyToOutputDirectory)
         {
             var elementList = new List<XElement>();
@@ -118,42 +118,6 @@ namespace dotnetCampus.SourceYard.Utils
             {
                 element.SetAttributeValue("CopyToOutputDirectory", "PreserveNewest");
             }
-        }
-
-        private List<string> GetFileList(string file)
-        {
-            if (string.IsNullOrEmpty(file) || !File.Exists(file))
-            {
-                return new List<string>();
-            }
-
-            var fileList = File.ReadAllLines(file).ToList();
-
-            fileList = RemoveTempFile(fileList);
-
-            return fileList;
-        }
-
-        private List<string> RemoveTempFile(List<string> fileList)
-        {
-            fileList.RemoveAll
-            (
-                temp => temp.StartsWith("obj\\")
-                        || temp.StartsWith("bin\\")
-            );
-
-            fileList.RemoveAll(temp =>
-            {
-                var pathRoot = Path.GetPathRoot(temp);
-                if (!string.IsNullOrEmpty(pathRoot))
-                {
-                    return temp.StartsWith(pathRoot);
-                }
-
-                return false;
-            });
-
-            return fileList;
         }
     }
 }

